In [1]:
from ultralytics import YOLO

# Load a model
model = YOLO("yolo11n-cls.pt")  # load a pretrained model (recommended for training)

dataset_path = "/mnt/d/tornetExamples/all_2014_and_2015"

# Train the model
results = model.train(data="/mnt/d/tornetExamples/all_2014_and_2015", augment=False, epochs=40, save=False, batch=4, freeze=7, imgsz=480)

New https://pypi.org/project/ultralytics/8.3.38 available 😃 Update with 'pip install -U ultralytics'
Ultralytics 8.3.36 🚀 Python-3.9.20 torch-2.5.1+cu124 CUDA:0 (NVIDIA GeForce RTX 3070, 8192MiB)
engine/trainer: task=classify, mode=train, model=yolo11n-cls.pt, data=/mnt/d/tornetExamples/all_2014_and_2015, epochs=40, time=None, patience=100, batch=4, imgsz=480, save=False, save_period=-1, cache=False, device=None, workers=8, project=None, name=train26, exist_ok=False, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=7, multi_scale=False, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, embed

E0000 00:00:1732788295.466534     107 cuda_dnn.cc:8310] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1732788295.480849     107 cuda_blas.cc:1418] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered


Overriding model.yaml nc=80 with nc=2

                   from  n    params  module                                       arguments                     
  0                  -1  1       464  ultralytics.nn.modules.conv.Conv             [3, 16, 3, 2]                 
  1                  -1  1      4672  ultralytics.nn.modules.conv.Conv             [16, 32, 3, 2]                
  2                  -1  1      6640  ultralytics.nn.modules.block.C3k2            [32, 64, 1, False, 0.25]      
  3                  -1  1     36992  ultralytics.nn.modules.conv.Conv             [64, 64, 3, 2]                
  4                  -1  1     26080  ultralytics.nn.modules.block.C3k2            [64, 128, 1, False, 0.25]     
  5                  -1  1    147712  ultralytics.nn.modules.conv.Conv             [128, 128, 3, 2]              
  6                  -1  1     87040  ultralytics.nn.modules.block.C3k2            [128, 128, 1, True]           
  7                  -1  1    295424  ultralytics

train: Scanning /mnt/d/tornetExamples/all_2014_and_2015/train... 37168 images, 0 corrupt: 100%|██████████| 37168/37168 [
val: Scanning /mnt/d/tornetExamples/all_2014_and_2015/test... 6448 images, 0 corrupt: 100%|██████████| 6448/6448 [00:00<


optimizer: 'optimizer=auto' found, ignoring 'lr0=0.01' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically... 
optimizer: SGD(lr=0.01, momentum=0.9) with parameter groups 39 weight(decay=0.0), 40 weight(decay=0.0005), 40 bias(decay=0.0)
TensorBoard: model graph visualization added ✅
Image sizes 480 train, 480 val
Using 8 dataloader workers
Logging results to runs/classify/train26
Starting training for 40 epochs...

      Epoch    GPU_mem       loss  Instances       Size


               classes   top1_acc   top5_acc: 100%|██████████| 806/806 [02:32<00:00,  5.29it/s]

                   all       0.89          1



      Epoch    GPU_mem       loss  Instances       Size


               classes   top1_acc   top5_acc: 100%|██████████| 806/806 [02:30<00:00,  5.37it/s]

                   all      0.893          1



      Epoch    GPU_mem       loss  Instances       Size


               classes   top1_acc   top5_acc: 100%|██████████| 806/806 [02:28<00:00,  5.41it/s]

                   all       0.89          1



      Epoch    GPU_mem       loss  Instances       Size


               classes   top1_acc   top5_acc: 100%|██████████| 806/806 [02:29<00:00,  5.38it/s]

                   all       0.89          1

      Epoch    GPU_mem       loss  Instances       Size



               classes   top1_acc   top5_acc: 100%|██████████| 806/806 [02:30<00:00,  5.35it/s]

                   all      0.894          1



      Epoch    GPU_mem       loss  Instances       Size


               classes   top1_acc   top5_acc: 100%|██████████| 806/806 [02:30<00:00,  5.36it/s]

                   all      0.894          1



      Epoch    GPU_mem       loss  Instances       Size


               classes   top1_acc   top5_acc: 100%|██████████| 806/806 [02:31<00:00,  5.33it/s]

                   all      0.893          1



      Epoch    GPU_mem       loss  Instances       Size


               classes   top1_acc   top5_acc: 100%|██████████| 806/806 [02:28<00:00,  5.45it/s]

                   all      0.898          1



      Epoch    GPU_mem       loss  Instances       Size


               classes   top1_acc   top5_acc: 100%|██████████| 806/806 [02:32<00:00,  5.29it/s]

                   all      0.902          1



      Epoch    GPU_mem       loss  Instances       Size


      10/40     0.185G      0.208          4        480: 100%|██████████| 9292/9292 [22:38<00:00,  6.84it/s]
               classes   top1_acc   top5_acc: 100%|██████████| 806/806 [02:25<00:00,  5.53it/s]

                   all      0.901          1



      Epoch    GPU_mem       loss  Instances       Size


      11/40     0.187G     0.2044          4        480: 100%|██████████| 9292/9292 [22:35<00:00,  6.86it/s]
               classes   top1_acc   top5_acc: 100%|██████████| 806/806 [02:29<00:00,  5.38it/s]

                   all      0.899          1

      Epoch    GPU_mem       loss  Instances       Size



      12/40     0.187G     0.2036          4        480: 100%|██████████| 9292/9292 [22:34<00:00,  6.86it/s]
               classes   top1_acc   top5_acc: 100%|██████████| 806/806 [02:28<00:00,  5.41it/s]

                   all      0.901          1



      Epoch    GPU_mem       loss  Instances       Size


      13/40     0.187G     0.2016          4        480: 100%|██████████| 9292/9292 [22:37<00:00,  6.84it/s]
               classes   top1_acc   top5_acc: 100%|██████████| 806/806 [02:29<00:00,  5.38it/s]

                   all      0.904          1

      Epoch    GPU_mem       loss  Instances       Size



      14/40     0.185G     0.1988          4        480: 100%|██████████| 9292/9292 [22:56<00:00,  6.75it/s]
               classes   top1_acc   top5_acc: 100%|██████████| 806/806 [02:28<00:00,  5.42it/s]

                   all      0.904          1



      Epoch    GPU_mem       loss  Instances       Size


      15/40     0.187G     0.1985          4        480: 100%|██████████| 9292/9292 [22:43<00:00,  6.82it/s]
               classes   top1_acc   top5_acc: 100%|██████████| 806/806 [02:29<00:00,  5.40it/s]

                   all      0.906          1



      Epoch    GPU_mem       loss  Instances       Size


      16/40     0.187G     0.1968          4        480: 100%|██████████| 9292/9292 [22:41<00:00,  6.82it/s]
               classes   top1_acc   top5_acc: 100%|██████████| 806/806 [02:27<00:00,  5.45it/s]

                   all      0.908          1



      Epoch    GPU_mem       loss  Instances       Size


      17/40     0.187G     0.1943          4        480: 100%|██████████| 9292/9292 [22:48<00:00,  6.79it/s]
               classes   top1_acc   top5_acc: 100%|██████████| 806/806 [02:27<00:00,  5.47it/s]

                   all      0.906          1



      Epoch    GPU_mem       loss  Instances       Size


      18/40     0.187G     0.1937          4        480: 100%|██████████| 9292/9292 [22:41<00:00,  6.83it/s]
               classes   top1_acc   top5_acc: 100%|██████████| 806/806 [02:34<00:00,  5.22it/s]

                   all      0.906          1



      Epoch    GPU_mem       loss  Instances       Size


      19/40     0.187G     0.1924          4        480: 100%|██████████| 9292/9292 [22:55<00:00,  6.76it/s]
               classes   top1_acc   top5_acc: 100%|██████████| 806/806 [02:29<00:00,  5.40it/s]

                   all      0.905          1



      Epoch    GPU_mem       loss  Instances       Size


      20/40     0.187G     0.1898          4        480: 100%|██████████| 9292/9292 [23:00<00:00,  6.73it/s]
               classes   top1_acc   top5_acc: 100%|██████████| 806/806 [02:27<00:00,  5.47it/s]

                   all      0.908          1



      Epoch    GPU_mem       loss  Instances       Size


      21/40     0.187G     0.1896          4        480: 100%|██████████| 9292/9292 [22:48<00:00,  6.79it/s]
               classes   top1_acc   top5_acc: 100%|██████████| 806/806 [02:27<00:00,  5.46it/s]

                   all      0.909          1



      Epoch    GPU_mem       loss  Instances       Size


      22/40     0.187G      0.189          4        480: 100%|██████████| 9292/9292 [22:21<00:00,  6.92it/s]
               classes   top1_acc   top5_acc: 100%|██████████| 806/806 [02:25<00:00,  5.53it/s]

                   all      0.908          1



      Epoch    GPU_mem       loss  Instances       Size


      23/40     0.187G     0.1871          4        480: 100%|██████████| 9292/9292 [22:35<00:00,  6.86it/s]
               classes   top1_acc   top5_acc: 100%|██████████| 806/806 [02:29<00:00,  5.39it/s]

                   all      0.907          1



      Epoch    GPU_mem       loss  Instances       Size


      24/40     0.187G     0.1859          4        480: 100%|██████████| 9292/9292 [22:36<00:00,  6.85it/s]
               classes   top1_acc   top5_acc: 100%|██████████| 806/806 [02:28<00:00,  5.44it/s]

                   all      0.906          1

      Epoch    GPU_mem       loss  Instances       Size



      25/40     0.187G     0.1846          4        480: 100%|██████████| 9292/9292 [22:40<00:00,  6.83it/s]
               classes   top1_acc   top5_acc: 100%|██████████| 806/806 [02:29<00:00,  5.38it/s]

                   all      0.906          1



      Epoch    GPU_mem       loss  Instances       Size


      26/40     0.185G     0.1827          4        480: 100%|██████████| 9292/9292 [23:02<00:00,  6.72it/s]
               classes   top1_acc   top5_acc: 100%|██████████| 806/806 [02:27<00:00,  5.46it/s]

                   all      0.905          1

      Epoch    GPU_mem       loss  Instances       Size



      27/40     0.187G     0.1823          4        480: 100%|██████████| 9292/9292 [22:34<00:00,  6.86it/s]
               classes   top1_acc   top5_acc: 100%|██████████| 806/806 [02:27<00:00,  5.45it/s]

                   all      0.904          1



      Epoch    GPU_mem       loss  Instances       Size


      28/40     0.187G     0.1773          4        480: 100%|██████████| 9292/9292 [22:39<00:00,  6.84it/s]
               classes   top1_acc   top5_acc: 100%|██████████| 806/806 [02:26<00:00,  5.49it/s]

                   all      0.904          1



      Epoch    GPU_mem       loss  Instances       Size


      29/40     0.187G     0.1781          4        480: 100%|██████████| 9292/9292 [22:42<00:00,  6.82it/s]
               classes   top1_acc   top5_acc: 100%|██████████| 806/806 [02:30<00:00,  5.34it/s]

                   all      0.903          1



      Epoch    GPU_mem       loss  Instances       Size


      30/40     0.185G     0.1774          4        480: 100%|██████████| 9292/9292 [22:41<00:00,  6.83it/s]
               classes   top1_acc   top5_acc: 100%|██████████| 806/806 [02:27<00:00,  5.46it/s]

                   all      0.904          1



      Epoch    GPU_mem       loss  Instances       Size


      31/40     0.187G     0.1751          4        480: 100%|██████████| 9292/9292 [22:51<00:00,  6.77it/s]
               classes   top1_acc   top5_acc: 100%|██████████| 806/806 [02:28<00:00,  5.43it/s]

                   all      0.904          1



      Epoch    GPU_mem       loss  Instances       Size


      32/40     0.187G     0.1712          4        480: 100%|██████████| 9292/9292 [22:43<00:00,  6.81it/s]
               classes   top1_acc   top5_acc: 100%|██████████| 806/806 [02:27<00:00,  5.48it/s]

                   all      0.903          1



      Epoch    GPU_mem       loss  Instances       Size


      33/40     0.187G     0.1712          4        480: 100%|██████████| 9292/9292 [22:44<00:00,  6.81it/s]
               classes   top1_acc   top5_acc: 100%|██████████| 806/806 [02:28<00:00,  5.43it/s]

                   all      0.902          1



      Epoch    GPU_mem       loss  Instances       Size


      34/40     0.187G     0.1698          4        480: 100%|██████████| 9292/9292 [22:36<00:00,  6.85it/s]
               classes   top1_acc   top5_acc: 100%|██████████| 806/806 [02:27<00:00,  5.45it/s]

                   all      0.901          1



      Epoch    GPU_mem       loss  Instances       Size


      35/40     0.187G     0.1675          4        480: 100%|██████████| 9292/9292 [23:07<00:00,  6.70it/s]
               classes   top1_acc   top5_acc: 100%|██████████| 806/806 [02:28<00:00,  5.41it/s]

                   all        0.9          1



      Epoch    GPU_mem       loss  Instances       Size


      36/40     0.187G     0.1658          4        480: 100%|██████████| 9292/9292 [22:41<00:00,  6.83it/s]
               classes   top1_acc   top5_acc: 100%|██████████| 806/806 [02:26<00:00,  5.51it/s]

                   all        0.9          1



      Epoch    GPU_mem       loss  Instances       Size


      37/40     0.187G     0.1617          4        480: 100%|██████████| 9292/9292 [22:33<00:00,  6.86it/s]
               classes   top1_acc   top5_acc: 100%|██████████| 806/806 [02:29<00:00,  5.38it/s]

                   all        0.9          1



      Epoch    GPU_mem       loss  Instances       Size


      38/40     0.185G     0.1624          4        480: 100%|██████████| 9292/9292 [22:49<00:00,  6.78it/s]
               classes   top1_acc   top5_acc: 100%|██████████| 806/806 [02:29<00:00,  5.38it/s]

                   all        0.9          1



      Epoch    GPU_mem       loss  Instances       Size


      39/40     0.187G     0.1585          4        480: 100%|██████████| 9292/9292 [22:46<00:00,  6.80it/s]
               classes   top1_acc   top5_acc: 100%|██████████| 806/806 [02:28<00:00,  5.44it/s]

                   all      0.898          1



      Epoch    GPU_mem       loss  Instances       Size


      40/40     0.187G     0.1562          4        480: 100%|██████████| 9292/9292 [22:49<00:00,  6.79it/s]
               classes   top1_acc   top5_acc: 100%|██████████| 806/806 [02:30<00:00,  5.34it/s]

                   all      0.899          1



40 epochs completed in 16.815 hours.
Optimizer stripped from runs/classify/train26/weights/last.pt, 3.2MB


Even on this attempt with more epochs (just to be certain that we don't increase again in accuracy later for some reason), we peak at .908 in accuracy, which is again far below the .94 (and nearly .95) marks that we were seeing in accuracy for our best mirroring-augmented model and the original TorNet model.

It seems likely that using the full set of 12 images as a concatenated image just isn't a great strategy here. There's a few things that could be going wrong with our transfer learning strategy:
- Without bounding boxes for the tornadoes, we can’t utilize YOLO object detection algorithms and are using a pre-trained classification model instead. We'd hoped that the sheer amount of data it had been trained on would have given it useful features to learn on here, but that doesn't seem to be the case so far.
- The 12 radar images were combined into single input images, and each "concatenation" concentrated individual radar plots into tiny patches - so the YOLO CNN had fewer pixels of interest to learn from. Maybe we could do a bit better with individual images instead (let's try that in the next notebook).